README.md
ModbusTCP
=========

We fork this library from ductsoup/ModbusTCP library, and fix some bugs on it (see the history.txt).

Visit the original ductsoup/ModbusTCP for details: https://github.com/ductsoup/ModbusTCP

Enjoy.


history.txt
2016/06/04
=============

1. Fixed: Line 36, add mb_server.begin() to initialize Ethernet.
2. A new implementation instead of word(*ptr++, *ptr++).

ModbusTCP.h
/*******************************************************************************
 * ModbusTCP.h
 *
 * Lightweight Modbus TCP 32-bit slave for Arduino. 
 * 
 * Three MODBUS function codes are supported; reading (0x04) and writing (0x10)
 * of holding registers and function 0x06 for compatiblilty. 
 * 
 * Three Modbus encodings are supported; UInt16, UInt32 and Float (Float and 
 * Double are identical on Arduino Uno). With these three primitive types you
 * can also support Int16, Int32, or packed bits (coils). 
 * 
 * Holding register addressing begins at 40001 and are contiguous.
 * 
 * Data is stored in native MODBUS UInt16 registers. To make it less error 
 * prone, get and set functions are included for each encoding that also do 
 * error checking. If you're crushed for resources feel free to comment out
 * the define statements and write to the registers directly.
 * 
 * Testing
 * This code was developed and tested using modpoll and Inductive Automation 
 * Ignition quick client running in trial mode.
 * 
 * http://www.modbusdriver.com/modpoll.html
 * https://www.inductiveautomation.com/scada-software
 * 
 * Credits
 * 
 * https://code.google.com/p/mudbus/
 * 
 * Thanks to Dee Wykoff and Martin Pettersson for making the Mudbus library 
 * available as a place to start.
 *
 * MIT License
 * Copyright (c) 2014 Ductsoup
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE. 
 */

/*******************************************************************************
 * 
 * MODBUS Reference
 * 
 * http://www.csimn.com/CSI_pages/Modbus101.html
 * 
 * Modicon convention notation
 * 
 * 0x = Coil = 00001-09999
 * 1x = Discrete Input = 10001-19999
 * 3x = Input Register = 30001-39999
 * 4x = Holding Register = 40001-49999
 * 
 * Modbus protocol defines a holding register as 16 bits wide; however, there is 
 * a widely used defacto standard for reading and writing data wider than 16 
 * bits. The most common are IEEE 754 floating point, and 32-bit integer. The 
 * convention may also be extended to double precision floating point and 64-bit 
 * integer data.
 * 
 * http://www.freemodbus.org/
 * 
 * ----------------------- MBAP Header --------------------------------------
 * 
 * <------------------------ MODBUS TCP/IP ADU(1) ------------------------->
 * <----------- MODBUS PDU (1') ---------------->
 * +-----------+---------------+------------------------------------------+
 * | TID | PID | Length | UID  |Code | Data                               |
 * +-----------+---------------+------------------------------------------+
 * |     |     |        |      |                                           
 * (2)   (3)   (4)      (5)    (6)                                          
 * (2)  ... MB_TCP_TID          = 0 (Transaction Identifier - 2 Byte) 
 * (3)  ... MB_TCP_PID          = 2 (Protocol Identifier - 2 Byte)
 * (4)  ... MB_TCP_LEN          = 4 (Number of bytes - 2 Byte)
 * (5)  ... MB_TCP_UID          = 6 (Unit Identifier - 1 Byte)
 * (6)  ... MB_TCP_FUNC         = 7 (Modbus Function Code)
 * 
 * (1)  ... Modbus TCP/IP Application Data Unit
 * (1') ... Modbus Protocol Data Unit
 * 
 *******************************************************************************/

#ifndef HEADER_ModbusTCP
#define HEADER_ModbusTCP

//
// Note: The Arduino IDE does not respect conditional included
// header files in the main sketch so you have to select your
// hardware there, and here.
//
#define MB_ETHERNET
//#define MB_CC3000

#ifdef MB_ETHERNET
#include <Ethernet.h>
#endif
#ifdef MB_CC3000
#include <Adafruit_CC3000.h>
#include <SPI.h>
  #if defined (_86DUINO)
  #define ADAFRUIT_CC3000_IRQ   42
  #define ADAFRUIT_CC3000_VBAT  44
  #define ADAFRUIT_CC3000_CS    43
  #elif defined (__AVR__) || defined (ARDUINO)
  #define ADAFRUIT_CC3000_IRQ   3
  #define ADAFRUIT_CC3000_VBAT  5
  #define ADAFRUIT_CC3000_CS    10
  #endif
#endif

//
// Application specific, adjust as necessary
//
#define MB_PORT 502
#define MB_REGISTERS_MAX 32
//
// Comment out the features you don't need                            
//
#define MB_Float            
#define MB_UInt32
#define MB_UInt16
//#define MB_DEBUG

class ModbusTCP {
public:
  ModbusTCP(void);
#ifdef MB_ETHERNET  
  void begin(uint8_t mac[6]);
#endif
#ifdef MB_CC3000  
  void begin(const char *ssid, const char *key, uint8_t secmode);
#endif
  void run(void);

#ifdef MB_Float
  int setFloat(uint16_t iAddress, float fValue);
  float getFloat(uint16_t iAddress);
#endif
#ifdef MB_UInt32    
  int setU32(uint16_t iAddress, uint32_t iValue);
  uint32_t getU32(uint16_t iAddress);
#endif
#ifdef MB_UInt16
  int setU16(uint16_t iAddress, uint16_t iValue);
  uint16_t getU16(uint16_t iAddress);
#endif
#ifdef MB_DEBUG
  void printHex(int num, int precision);
  void printMB(char *s, int n);
#endif
private:
  uint16_t mb_reg[MB_REGISTERS_MAX];
  uint8_t mb_adu[260];
#ifdef MB_ETHERNET
    EthernetServer mb_server;
#endif
#ifdef MB_CC3000
    Adafruit_CC3000 mb_client;
    Adafruit_CC3000_Server mb_server;
#endif  
};
//
// MODBUS Function Codes
//
#define MB_FC_NONE 0
#define MB_FC_READ_COILS 1
#define MB_FC_READ_DISCRETE_INPUT 2
#define MB_FC_READ_REGISTERS 3 // implemented
#define MB_FC_READ_INPUT_REGISTERS 4
#define MB_FC_WRITE_COIL 5
#define MB_FC_WRITE_REGISTER 6 // implemented
#define MB_FC_WRITE_MULTIPLE_COILS 15
#define MB_FC_WRITE_MULTIPLE_REGISTERS 16 // implemented
//
// MODBUS Error Codes
//
#define MB_EC_NONE 0
#define MB_EC_ILLEGAL_FUNCTION 1
#define MB_EC_ILLEGAL_DATA_ADDRESS 2
#define MB_EC_ILLEGAL_DATA_VALUE 3
#define MB_EC_SLAVE_DEVICE_FAILURE 4
//
// MODBUS MBAP offsets
//
#define MB_TCP_TID          0 // ignored
#define MB_TCP_PID          2 // ignored
#define MB_TCP_LEN          4
#define MB_TCP_UID          6 // ignored
#define MB_TCP_FUNC         7
#define MB_TCP_DATA         8

#endif


ModbusTCP.cpp
/*******************************************************************************
 * ModbusTCP.cpp
 */
#include <Arduino.h>
#include "ModbusTCP.h"

ModbusTCP::ModbusTCP(void) : 
#ifdef MB_ETHERNET
mb_server(MB_PORT) 
#endif
#ifdef MB_CC3000
mb_client(ADAFRUIT_CC3000_CS, ADAFRUIT_CC3000_IRQ, ADAFRUIT_CC3000_VBAT, SPI_CLOCK_DIVIDER),
mb_server(MB_PORT)
#endif
{
}
#ifdef MB_ETHERNET
void ModbusTCP::begin(uint8_t mac[6]){
  
  Serial.print(F("Requesting DHCP ..."));
  if (Ethernet.begin(mac) == 0) {
    Serial.println("failed");
    while (1);
  }
  else
    Serial.println();

  Serial.print(F("Listening on "));
  for (byte thisByte = 0; thisByte < 4; thisByte++) {
    Serial.print(Ethernet.localIP()[thisByte], DEC);
    Serial.print(thisByte < 3 ? "." : "");
  }
  Serial.print(":");
  Serial.print(MB_PORT);
  Serial.println(" ...");
  mb_server.begin(); // add this function to begin
}
#endif
#ifdef MB_CC3000
void ModbusTCP::begin(const char *ssid, const char *key, uint8_t secmode){

  mb_client.begin();
  Serial.print(F("Connecting to ")); Serial.print(ssid); Serial.println(F(" ..."));
  mb_client.connectToAP(ssid, key, secmode); 
  while (!mb_client.checkDHCP())
      delay(100);
  uint32_t ipAddress, netmask, gateway, dhcpserv, dnsserv;
  mb_client.getIPAddress(&ipAddress, &netmask, &gateway, &dhcpserv, &dnsserv);
  Serial.print(F("Listening on "));
  mb_client.printIPdotsRev(ipAddress);
  Serial.print(F(":")); 
  Serial.println(MB_PORT);
  mb_server.begin();
}
#endif

void ModbusTCP::run(void){

  int16_t i, iStart, iQty;
  int16_t iTXLen = 0;       // response packet length
  uint8_t *ptr, iFC = MB_FC_NONE, iEC = MB_EC_NONE;
  int16_t tmp;
  //  
  // Initialize and check for a request from a MODBUS master
  //
#ifdef MB_ETHERNET
  EthernetClient clientrequest = mb_server.available();
#endif
#ifdef MB_CC3000
  Adafruit_CC3000_ClientRef clientrequest = mb_server.available();
#endif
  if(clientrequest.available()) {
    //
    // Retrieve request
    //
    for (i = 0 ; clientrequest.available() ; i++) 
      mb_adu[i] = clientrequest.read();
#ifdef MB_DEBUG   
    printMB("RX: ", word(mb_adu[MB_TCP_LEN],mb_adu[MB_TCP_LEN+1])+6);  
#endif    
    //
    // Unpack the function code
    //
    iFC = mb_adu[MB_TCP_FUNC];
  }
  //
  // Handle request
  //
  switch(iFC) {
  case MB_FC_NONE:
    break;
  case MB_FC_READ_REGISTERS: 
    //
    // 03 (0x03) Read Holding Registers
    //
    // modpoll -m tcp -t 4:float -r 40001 -c 1 -1 192.168.x.x
    //
    //     [TransID] [ProtID-] [Length-] [Un] [FC] [Start--] [Qty----] 
    // RX: 0x00 0x01 0x00 0x00 0x00 0x06 0x01 0x03 0x9C 0x40 0x00 0x02 
    //
    //     [TransID] [ProtID-] [Length-] [Un] [FC] [Bc] [float------------] 
    // TX: 0x00 0x01 0x00 0x00 0x00 0x07 0x01 0x03 0x04 0x20 0x00 0x47 0xF1
    //
    // 123456.0 = 0x00 0x20 0xF1 0x47 (IEEE 754)
    //
    // Unpack the start and length
    //
    ptr = mb_adu + MB_TCP_DATA;
	tmp = *ptr++;
    iStart = word(tmp, *ptr++) - 40000;
	tmp = *ptr++;
    iQty = 2*word(tmp, *ptr);
    //
    // check for valid register addresses     
    //
    if (iStart < 0 || (iStart + iQty/2 - 1) > MB_REGISTERS_MAX) {
      iEC = MB_EC_ILLEGAL_DATA_ADDRESS;
      break;
    }
    //
    // Write data length
    //
    ptr = mb_adu + MB_TCP_DATA;
    *ptr++ = iQty;
    //
    // Write data
    //
    for (i = 0 ; i < iQty/2 ; i++) {
      *ptr++ = highByte(mb_reg[iStart + i]);
      *ptr++ =  lowByte(mb_reg[iStart + i]);
    }
    iTXLen = iQty + 9;
    break;

  case MB_FC_WRITE_REGISTER:
    //
    // 06 (0x06) Write register 
    //
    ptr = mb_adu + MB_TCP_DATA;
    tmp = *ptr++;	
    iStart = word(tmp, *ptr++) - 40000;
    //
    // check for valid register addresses     
    //   
    if (iStart < 0 || (iStart - 1) > MB_REGISTERS_MAX) {
      iEC = MB_EC_ILLEGAL_DATA_ADDRESS;
      break;
    }      
    // Unpack and store data
    //
	tmp = *ptr++;
    mb_reg[iStart] = word(tmp, *ptr);
    //
    // Build a response 
    //
    iTXLen = 12;             
    break;

  case MB_FC_WRITE_MULTIPLE_REGISTERS: 
    //
    // 16 (0x10) Write Multiple registers
    //
    // modpoll -m tcp -t 4:float -r 40001 -c 1 -1 192.168.x.x 123456.0
    //
    //     [TransID] [ProtID-] [Length-] [Un] [FC] [Start--] [Qty----] [Bc] [float------------] 
    // RX: 0x00 0x01 0x00 0x00 0x00 0x0B 0x01 0x10 0x9C 0x40 0x00 0x02 0x04 0x20 0x00 0x47 0xF1
    //
    // 123456.0 = 0x00 0x20 0xF1 0x47 (IEEE 754)
    //
    // Unpack the start and length
    //
    ptr = mb_adu + MB_TCP_DATA;
	tmp = *ptr++;	
    iStart = word(tmp, *ptr++) - 40000;
	tmp = *ptr++;
    iQty = 2*word(tmp, *ptr);      
    //
    // check for valid register addresses     
    //   
    if (iStart < 0 || (iStart + iQty/2 - 1) > MB_REGISTERS_MAX) {
      iEC = MB_EC_ILLEGAL_DATA_ADDRESS;
      break;
    }
    //
    // Unpack and store data
    //
    ptr = mb_adu + MB_TCP_DATA+5;
    // todo: check for valid length
    for (i = 0 ; i < iQty/2 ; i++) {
	  tmp = *ptr++;
      mb_reg[iStart + i] = word(tmp, *ptr++);
    }
    //
    // Build a response 
    //
    iTXLen = 12;
    break;

  default:
    iEC = MB_EC_ILLEGAL_FUNCTION;
    break;    
  }
  //
  // Build exception response if necessary because we were too
  // lazy to do it earlier. Other responses should already be
  // built.
  //
  if (iEC) {
    ptr = mb_adu + MB_TCP_FUNC;
    *ptr = *ptr++ | 0x80;        // flag the function code
    *ptr = iEC;                  // write the exception code
    iTXLen = 9;
  }
  //
  // If there's a response, transmit it
  //
  if (iFC) {
    ptr = mb_adu + MB_TCP_LEN;    // write the header length
    *ptr++ = 0x00;
    *ptr = iTXLen - MB_TCP_UID;  
    clientrequest.write(mb_adu, iTXLen); // send it        
#ifdef MB_DEBUG   
    printMB("TX: ", word(mb_adu[MB_TCP_LEN], mb_adu[MB_TCP_LEN+1]) + MB_TCP_UID);  
#endif                                                
  }
}

#ifdef MB_Float
int ModbusTCP::setFloat(uint16_t iAddress, float fValue) {

  int iRet; 
  union {
    float f;
    uint16_t w[2];
  } 
  fw;

  for (iRet = 0 ; !iRet ; iRet = 2) {
    if ((iAddress -= 40001) < 0 || (iAddress + 1) >= MB_REGISTERS_MAX) // invalid address
      break;
    fw.f = fValue;
    mb_reg[iAddress++] = fw.w[0];
    mb_reg[iAddress] = fw.w[1]; 
  }
  return(iRet);  
}
float ModbusTCP::getFloat(uint16_t iAddress) {

  union {
    float fRet;
    uint16_t w[2];
  } 
  fw;

  if ((iAddress -= 40001) < 0 || (iAddress + 1) >= MB_REGISTERS_MAX) // invalid address
    fw.fRet = NULL;
  else {
    fw.w[0] = mb_reg[iAddress++];  
    fw.w[1] = mb_reg[iAddress];    
  }
  return(fw.fRet);  
}
#endif
#ifdef MB_UInt32
int ModbusTCP::setU32(uint16_t iAddress, uint32_t iValue) {

  int iRet; 
  for (iRet = 0 ; !iRet ; iRet = 2) {
    if ((iAddress -= 40001) < 0 || (iAddress + 1) >= MB_REGISTERS_MAX) // invalid address
      break;
    mb_reg[iAddress++] = iValue & 0xffff;
    mb_reg[iAddress] = iValue >> 16;
  }
  return(iRet);  
}
uint32_t ModbusTCP::getU32(uint16_t iAddress) {

  uint32_t lRet; 

  if ((iAddress -= 40001) < 0 || (iAddress + 1) >= MB_REGISTERS_MAX) // invalid address
    lRet = NULL;
  else {
    lRet = (long(mb_reg[iAddress++]) | (long(mb_reg[iAddress+1]) << 16));
  }
  return(lRet);  
}
#endif
#ifdef MB_UInt16
int ModbusTCP::setU16(uint16_t iAddress, uint16_t iValue) {

  int iRet; 
  for (iRet = 0 ; !iRet ; iRet = 1) {
    if ((iAddress -= 40001) < 0 || (iAddress >= MB_REGISTERS_MAX)) // invalid address
      break;
    mb_reg[iAddress] = iValue;
  }
  return(iRet);  
}
uint16_t ModbusTCP::getU16(uint16_t iAddress) {

  uint16_t iRet; 

  if ((iAddress -= 40001) < 0 || (iAddress >= MB_REGISTERS_MAX)) // invalid address
    iRet = NULL;
  else
    iRet = mb_reg[iAddress];

  return(iRet);  
}
#endif
#ifdef MB_DEBUG
void ModbusTCP::printHex(int num, int precision) {

  char tmp[16];
  char format[128];

  sprintf(format, "0x%%.%dX ", precision);

  sprintf(tmp, format, num);
  Serial.print(tmp);
}
void ModbusTCP::printMB(char *s, int n) {

  int i;

  Serial.print(s);
  for (i = 0 ; i < n ; i++) {
    printHex(mb_adu[i], 2); 
    if (i == 6) Serial.print("- ");
    if (i == 7) Serial.print("- ");      
  }
  Serial.println("");      
}
#endif



